sources/engine/Xenko.Voxels/Voxels/GraphicsCompositor/VoxelRenderer.cs

```diff
@@ -0,0 +1,260 @@
+                if (dataVolume.VoxelGridSnapping)
+                {
+                    matTrans /= storageContext.RealVoxelSize();
+                    matTrans.X = (float)Math.Floor(matTrans.X);
+                    matTrans.Y = (float)Math.Floor(matTrans.Y);
+                    matTrans.Z = (float)Math.Floor(matTrans.Z);
+                    matTrans *= storageContext.RealVoxelSize();
+
+                    corMatrix = Matrix.Scaling(matScale) * Matrix.Translation(matTrans);
+                    storageContext.Matrix = corMatrix;
+                }
+                storageContext.Translation = matTrans;
+                storageContext.VoxelSpaceTranslation = matTrans / storageContext.RealVoxelSize();
+
+                //Update storage
+                dataVolume.Storage.UpdateFromContext(storageContext);
+
+                //Transfer voxelization info
+                processedVolume.VisualizeVoxels = dataVolume.VisualizeVoxels;
+                processedVolume.Storage = dataVolume.Storage;
+                processedVolume.StorageContext = storageContext;
+                processedVolume.VoxelizationMethod = dataVolume.VoxelizationMethod;
+                processedVolume.VoxelVisualization = dataVolume.VoxelVisualization;
+                processedVolume.VisualizationAttribute = dataVolume.VisualizationAttribute;
+                processedVolume.Voxelize = dataVolume.Voxelize;
+                processedVolume.OutputAttributes = dataVolume.Attributes;
+
+                processedVolume.Attributes.Clear();
+                processedVolume.passList.Clear();
+                processedVolume.groupedPasses.Clear();
+
+                processedVolume.passList.defaultVoxelizationMethod = dataVolume.VoxelizationMethod;
+
+
+                //Create final list of attributes (including temporary ones)
+                foreach (var attr in dataVolume.Attributes)
+                {
+                    attr.CollectAttributes(processedVolume.Attributes, VoxelizationStage.Initial, true);
+                }
+
+                //Allocate textures and space in the temporary buffer
+                foreach (var attr in processedVolume.Attributes)
+                {
+                    attr.Attribute.PrepareLocalStorage(storageContext, dataVolume.Storage);
+                    if (attr.Output)
+                    {
+                        attr.Attribute.PrepareOutputStorage(storageContext, dataVolume.Storage);
+                    }
+                    else
+                    {
+                        attr.Attribute.ClearOutputStorage();
+                    }
+                }
+                dataVolume.Storage.UpdateTempStorage(storageContext);
+
+                //Create list of voxelization passes that need to be done
+                dataVolume.Storage.CollectVoxelizationPasses(processedVolume, storageContext);
+
+                //Group voxelization passes where the RenderStage can be shared
+                //TODO: Group identical attributes
```

Comment:
If a user adds two of the same attribute and sets them up identically, ideally only one would be voxelized and the other would share its data; that currently doesn't happen. That situation might sound artificial, but since attributes can be added to the pipeline automatically by modifiers (like the solidify modifier), if the user adds their own Solidity output attribute (since they can't otherwise access the solidity data), there will be duplicate processing and storage. 

Still its not a common occurrence and there needs to be some restructuring of how attributes allocate their textures and buffer space to have this work, so I've left it for now. Hopefully that makes sense :)